io-engine/src/bdev/nexus/nexus_persistence.rs

```diff
@@ -171,51 +173,45 @@ impl<'n> Nexus<'n> {
                 nexus_info.clean_shutdown = true;
             }
         }
-        self.save(&persistent_nexus_info).await;
+
+        self.save(&persistent_nexus_info).await
     }
 
-    // Save the nexus info to the store. This is integral to ensuring data
+    // Saves the nexus info to the store. This is integral to ensuring data
     // consistency across restarts of Mayastor. Therefore, keep retrying
     // until successful.
-    // TODO: Should we give up retrying eventually?
-    async fn save(&self, info: &PersistentNexusInfo) {
-        let mut output_err = true;
-        let nexus_uuid = self.uuid().to_string();
-        // If a key has been provided use this to store the NexusInfo.
-        // If a key is not provided, use the nexus uuid as the key.
+    async fn save(&self, info: &PersistentNexusInfo) -> Result<(), Error> {
+        // If a key has been provided, use it to store the NexusInfo; use the
+        // nexus uuid as the key otherwise.
         let key = match &info.key {
             Some(k) => k.clone(),
             None => self.uuid().to_string(),
         };
 
+        let mut retry = PersistentStore::retries();
         loop {
-            match PersistentStore::put(&key, &info.inner).await {
-                Ok(_) => {
-                    // The state was saved successfully.
-                    break;
-                }
-                Err(e) => {
-                    // Output an error message on first failure. Thereafter
-                    // silently retry.
-                    if output_err {
-                        error!(
-                            "Failed to persist nexus information for nexus {}, UUID {} with error {}. Retrying...",
-                            self.name,
-                            nexus_uuid,
-                            e
-                        );
-                        output_err = false;
-                    }
+            let Err(err) = PersistentStore::put(&key, &info.inner).await else {
+                trace!(?key, "{self:?}: the state was saved successfully");
+                return Ok(());
+            };
+
+            retry -= 1;
+            if retry == 0 {
```

Comment:
should this check be <= 0 ? What if retries starts with 0 and then we decrement it.

Reply:
Also making me wonder if this should be time based rather than retry based.
Example, what if endpoint rejecting immediately due to some temporary condition? Then we only sleep for 1s between retries, so with 5 retries for example, we'd give up after 5s which seems very small?

Reply:
I made retries tunable with CLI args, so an admin can tune it to a particular setup/network.